### ISSUE TITLE
PSModulePath for Pash

### ISSUE THREAD

**Issue by sburnicki** [Pash-Project/Pash#332]
I'd like to discuss what we do with the PSModulePath in Pash. Obviously, this environment variable describes locations where modules can be located, s.t. Powershell can find them.

I want to implement this in Pash, so the new module support (#331) will also work for installed modules.

Here are some issues:
- Naming: Should we use the same environment variable name as Powershell, i.e. PSModulePath? While this would be transparent for users, it could cause trouble on windows, as it's the same variable as Powershell uses.
- Default values: We should make up our minds where modules should be by default. E.g. on Linux I would suggest `/usr/local/Pash/modules/` and `~/.Pash/modules/`. I don't know what is common on Mac. Any suggestions? What about windows?

**Comment by sburnicki** [comment]
Yes, I know, that's why I'd tend to use PSModulePath on all platforms. Anyway, I'm not really sure how to deal with the default value of that variable...
While I think `~/.Pash/modules/` should be fine for Linux users, it doesn't solve system wide installations, a topic we didn't discuss at all, yet.
Maybe these default values should be stored in some kind of config file, so package maintainers can define them one day.

**Comment by Jaykul** [comment]
I'd say PSModulePath makes sense *if* you don't think that having all the default PowerShell modules in your PSModulePath is going to cause problems. The goal is full compatibility, but realistically, what percentage of Microsoft's built-in system modules do you think will **never** work in Pash on Windows? 

If trying to load PowerShell core modules results in errors, you're going to wish you'd picked a different name if you ever decide to implement Module autoloading in the future.  

It might be better to pick a new location, and then add your paths to PSModulePath so that it can pick up your modules ;-)  after all, the only problem I see with NOT using PSModulePath is that installers may modify PSModulePath, but not PashModulePath (or whatever), right? Is it a big deal to have a second variable?

**Comment by Jaykul** [comment]
Randomly: A couple of other thoughts:

As a hard-core PowerShell geek, I find it weird to think you're going to put Modules in a hidden folder on Linux.  Almost as weird as I thought it was when I first found they put them out in the open in "Documents" on Windows.

I think `~/.Pash/modules` would work on Mac too, but you could use /Library/Pash/Modules and ~/Library/Pash/Modules instead I think ... maybe take a look at how zsh and it's users are doing things like this?

I definitely suggest you 3 paths like we have in PS4 now: 
1. The system or shell install path, (C:\Windows\System32\...) 
2. The user's documents, (~\Documents\Pash or ~/.Pash or ~/Library/Pash)
3. The new machine-local path in C:\Program Files\ (or /Library or /etc/).

1 and 3 are confusing for Pash on windows, because they might be the same root folder (C:\Program Files\Pash\), so maybe you want to think harder about that.

P.S. It's amusing to me that we have three default module locations, but only 2 default profile locations. They didn't think of having a computer-wide common modules folder at the beginning (they figured people would just install in ProgramFiles\MyApp\Modules and append to the PSModulePath) and they didn't want to add another profile script when they decided that PowerShellGet means they needed the `C:\Program Files\WindowsPowerShell\Modules` folder in PS4 ;-)

**Comment by ForNeVeR** [comment]
About usage of `Program Files` on Windows - I'd prefer if no program requires writing to this directory. I think we should prefer `ProgramData` directory (as Chocolatey does, for example).

**Comment by Jaykul** [comment]
First of all, the `Program Files\WindowsPowerShell\Modules` location is hard-coded in PowerShell -- I'm simply suggesting Pash follow suit if it creates it's own PashModulePath variable.

Having said that: Modules aren't program data, they're programs. 

ProgramData is where modules should store machine-level configuration, not where the modules should be stored.  The whole point of Program Files is that it requires root/administrator access to install *or alter* files there.  You would still have your user profile location for installing, so you're never *required* to write to that directory.

P.S. Chocolatey's use of `ProgramData` is an edge case (they download to there, and store the installers there, but most of the packages install their application via MSI installers into `Program Files`).  Chocolatey does (incorrectly, imho) store it's apps there, **but they compensate** by locking their folder to `BuiltIn\Administrators` -- the result is that it behaves (somewhat) like `Program Files`  -- and incidentally, that decision results in some **rather annoying** behavior: you have to to be elevated (aka root or sudo root) in order to install things with chocolatey, *even if you're* ***not*** *installing to `Program Files`*, because their config files all require admin rights to edit, as does their download folder.

**Comment by sburnicki** [comment]
@Jaykul: Thanks for all your comments. As I'm basically a Linux guy, I'm certainly missing some of these insights into Powershell and Windows.

First of all: I think you are right that we should start with a `PashModulePath`, as it will avoid lots of trouble and keeps both products separate. Mac/Unix users won't care at all, I think, as they aren't used to any `PSModulePath` on their system anyway.

Concerning it's value on Linux:
Most applications (also other shells) have their own hidden folder in the user's home directory. They use it to store internal application specific data, but also addons. It's pretty common on Linux that you need to install your addons to such a hidden folder.
For me it feels weird that modules are installed under `Documents` on Windows. After all they are programs, as you said. I always thought of `Documents` as a location to save my personal documents to. And applications/application extensions don't count as personal documents for me ;-)

For the machine-local stuff I suggest `/usr/lib/Pash/` for modules installed with Pash, and `/usr/share/Pash/modules` for machine wide additionally installed modules. This should also fit most other application's behavior (take python as an example). However, we need a way to make that configurable, because the distro's package maintainer should have the last word on where Pash is installed.

Concerning Mac:
As for example zsh is just a Unix/Linux application run on OS X, it simply behaves the same way as on Linux. Is any Mac guy here to tell me what a solution would be *that feels native*?

Concerning Windows:
I think we should probably start off with `~\Documents\Pash` and `%programfiles%\Pash`. We could think about subdirectories in the latter one to separate things that shipped with Pash from newly installed modules.

**Comment by Jaykul** [comment]
You know, after thinking about it ... maybe you don't really need to distinguish between "modules that shipped with Pash" and other modules that _an administrator_ wants to install. That is, this is Open Source, not Microsoft, and maybe you can just leave it up to users to not break their installs, and have only two locations: system and user.

I was just discussing with OneGet team members the fact that they have an awkward problem because they are shipping a module as part of the PowerShell installation (or part of the OS) that they want to (be able to) update later...

It's an unlikely problem for Pash, but it highlights the inherent issues in making a one set of modules special ;-)

Of course, Microsoft made the distinction into a real problem because when autoloading (e.g. running a command without first importing the module containing that command) PowerShell prefers commands from system modules *regardless* of the PSModulePath order -- presumably you won't make that mistake ;-)

**Comment by ArsenShnurkov** [comment]
It is easy to determine current OS - http://stackoverflow.com/a/5117005/1709408

I propose to use the same variable on non-windows platforms (PSModulePath),
but on windows to use special one - PashModulePath.

This will cause less confusion for linux users, i think.

**Comment by ArsenShnurkov** [comment]
~/bin/ folder is used in Suse and Ubuntu:
http://www.linuxforums.org/forum/suse-linux/19974-what-bin-folder-home-dir.html

make install will probably put binaries in a place like $(prefix)/bin
http://stackoverflow.com/questions/2139854/automake-generating-binaries-to-bin-instead-of-in-src

so i propose ~/bin/Pash/modules for user modules
and /usr/lib/Pash/modules for distribution-installed modules

**Comment by sburnicki** [comment]
Thanks for this investiagtion. These paths make sense for Ubuntu. In general we will need to make this configurable, so it can be set during packaging for different platforms and distributions.

**Comment by Phrohdoh** [comment]
Under OS X `~/.pash/` would be perfectly fine and expected.

**Comment by jpumc** [comment]
I suggest **~/.local/lib/pash/modules** for user modules on Linux.
